File: AR7ExportToolbar/ToolStripExport/ExportToolStripButton.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using AR7ExportToolbar.ToolStripExport.Exporters;
using GrapeCity.ActiveReports.Document;
using GrapeCity.ActiveReports.Viewer.Win;

namespace AR7ExportToolbar.ToolStripExport
{
	public class ExportToolStripButton : ToolStripDropDownButton
	{
		private static readonly Image ToolStripButtonImage = new Bitmap(typeof(ExportToolStripButton), "ReportViewer.bmp");

		private readonly Viewer host;

		public IDictionary<string, ExportItem> AvailableExports { get; private set; }
		public IDictionary<string, ExportItem> VerifiedAvailableExports { get; private set; }

		private Dictionary<string, ToolStripMenuItem> dropDownItems = null;

		private bool dropDownItemsCreated = false;

		public ExportToolStripButton(Viewer host)
			: base("Export")
		{
			this.host = host;

			AvailableExports = new Dictionary<string, ExportItem>();

			AddAllExports();

			this.DropDownOpening += OnDropDownOpening;
		}

		private void AddAllExports()
		{
			var item = new ExportItem();
			item.Name = "PDF";
			item.LocalizedName = "PDF";
			item.PageReportExport = new PageReportExporter("GrapeCity.ActiveReports.Export.Pdf.Page.PdfRenderingExtension", "GrapeCity.ActiveReports.Export.Pdf.v7");
			item.SectionReportExport = new PdfSectionReportExporter();

			AvailableExports.Add("PDF", item);
		}

		private void VerifyExports()
		{
			var verifiedExports = new Dictionary<string, ExportItem>();

			foreach (var exportKV in AvailableExports)
			{
				var item = exportKV.Value;
				if (item != null)
				{
					bool available = false;

					if (item.PageReportExport != null && item.PageReportExport.IsAvailable())
					{
						available = true;
					}

					if (item.SectionReportExport != null && item.SectionReportExport.IsAvailable())
					{
						available = true;
					}

					if (available)
					{
						verifiedExports.Add(exportKV.Key, exportKV.Value);
					}
				}
			}
		}

		private void CreateDropDownItems()
		{
			VerifyExports();

			dropDownItems = new Dictionary<string, ToolStripMenuItem>();

			foreach (var export in VerifiedAvailableExports)
			{
				var menuItem = new ToolStripMenuItem(export.Value.LocalizedName, ToolStripButtonImage, OnExportItemClicked);
				
				dropDownItems.Add(export.Key, menuItem);
				DropDownItems.Add(menuItem);
			}

			dropDownItemsCreated = true;
		}

		private void OnDropDownOpening(object sender, EventArgs e)
		{
			if (!dropDownItemsCreated)
			{
				CreateDropDownItems();
			}

			
		}

		private void ToggleDropDownItemVisibility(SectionDocument doc)
		{
		}

		private void OnExportItemClicked(object sender, EventArgs e)
		{
			var item = sender as ToolStripItem;
		}

		protected override void OnDropDownOpened(EventArgs e)
		{
			base.OnDropDownOpened(e);
		}
	}
}

File: AR7ExportToolbarVB/ToolStripExport/ExportToolStripButton.vb
Imports GrapeCity.ActiveReports.Viewer.Win

Public Class ExportToolStripButton
	Inherits ToolStripDropDownButton

	Private Shared ReadOnly ToolStripButtonImage As Bitmap = New Bitmap(GetType(ExportToolStripButton), "ReportViewer.bmp")

	Private ReadOnly _host As Viewer

	Private _availableExports As IDictionary(Of String, ExportItem)
	Public Property AvailableExports As IDictionary(Of String, ExportItem)
		Get
			Return _availableExports
		End Get
		Private Set(value As IDictionary(Of String, ExportItem))
			_availableExports = value
		End Set
	End Property

	Private _verifiedAvailableExports As IDictionary(Of String, ExportItem)
	Public Property VerifiedAvailableExports As IDictionary(Of String, ExportItem)
		Get
			Return _verifiedAvailableExports
		End Get
		Private Set(value As IDictionary(Of String, ExportItem))
			_verifiedAvailableExports = value
		End Set
	End Property

	Private _dropDownItems As Dictionary(Of String, ToolStripMenuItem) = Nothing

	Private _dropDownItemsCreated As Boolean = False

	Public Sub New(host As Viewer)
		MyBase.New("Export")

		_host = host

		Enabled = False
		AttachHostEvents()

		AvailableExports = New Dictionary(Of String, ExportItem)()
		AddAllExports()

		Image = ToolStripButtonImage
		ImageTransparentColor = ToolStripButtonImage.GetPixel(0, 0)

		AddHandler Me.DropDownOpening, AddressOf OnDropDownOpening

	End Sub

	Protected Overrides Sub Dispose(disposing As Boolean)
		If disposing Then
			DetachHostEvents()
		End If

		MyBase.Dispose(disposing)
	End Sub

	Private Sub AttachHostEvents()
		AddHandler _host.LoadCompleted, AddressOf OnHostLoadCompleted
	End Sub

	Private Sub DetachHostEvents()
		RemoveHandler _host.LoadCompleted, AddressOf OnHostLoadCompleted
	End Sub

	Private Sub OnHostLoadCompleted(sender As Object, e As EventArgs)
		Enabled = True
	End Sub

	Private Sub AddAllExports()
		AddAvailableExport("PDF", "PDF", New PdfSectionReportExporter())
	End Sub

	Private Sub AddAvailableExport(name As String, localizedName As String, exporter As ISectionReportExporter)
		Dim item = New ExportItem()
		With item
			.Name = name
			.LocalizedName = localizedName
			.SectionReportExport = exporter
		End With

		AvailableExports.Add(name, item)
	End Sub

	Private Sub VerifyExports()
		Dim verifiedExports = New Dictionary(Of String, ExportItem)()

		For Each exportKV In AvailableExports
			Dim item = exportKV.Value

			If Not (item Is Nothing) Then
				Dim available = False

				If (Not (item.SectionReportExport Is Nothing)) AndAlso (item.SectionReportExport.IsAvailable()) Then
					available = True
				End If

				If available Then
					verifiedExports.Add(exportKV.Key, exportKV.Value)
				End If
			End If
		Next

		VerifiedAvailableExports = verifiedExports
	End Sub

	Private Sub CreateDropDownItems()
		VerifyExports()

		_dropDownItems = New Dictionary(Of String, ToolStripMenuItem)()

		For Each export In VerifiedAvailableExports
			Dim menuItem As ToolStripMenuItem = New ToolStripMenuItem(export.Value.LocalizedName, ToolStripButtonImage, AddressOf OnExportItemClicked)
			menuItem.Tag = export.Value
			menuItem.ImageTransparentColor = ToolStripButtonImage.GetPixel(0, 0)

			_dropDownItems.Add(export.Key, menuItem)
			DropDownItems.Add(menuItem)
		Next

		_dropDownItemsCreated = True
	End Sub

	Private Sub OnDropDownOpening(sender As Object, e As EventArgs)
		If Not _dropDownItemsCreated Then
			CreateDropDownItems()
		End If
	End Sub

	Private Sub OnExportItemClicked(sender As Object, e As EventArgs)
		Dim item As ToolStripItem = TryCast(sender, ToolStripItem)

		If item Is Nothing Then Return

		Dim exportItem As ExportItem = TryCast(item.Tag, ExportItem)

		If exportItem Is Nothing Then Return

		Dim export As ISectionReportExporter = exportItem.SectionReportExport
		If export Is Nothing Then Return

		Dim exporter As GrapeCity.ActiveReports.Export.IDocumentExportEx = export.GetExporter()
		If exporter Is Nothing Then Return

		Using saveAsDlg As SaveFileDialog = New SaveFileDialog()
			saveAsDlg.DefaultExt = export.DefaultExtension
			saveAsDlg.Filter = export.FileDialogFilter

			If Not saveAsDlg.ShowDialog(_host) = DialogResult.OK Then Return

			_host.Export(exporter, New IO.FileInfo(saveAsDlg.FileName))
		End Using

	End Sub

End Class
